app/src/main/java/org/groundplatform/android/ui/datacollection/DataCollectionHelper.kt

```diff
@@ -0,0 +1,128 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.groundplatform.android.ui.datacollection
+
+import androidx.lifecycle.SavedStateHandle
+import javax.inject.Inject
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.withTimeoutOrNull
+import org.groundplatform.android.model.Survey
+import org.groundplatform.android.model.job.Job
+import org.groundplatform.android.model.task.Task
+import org.groundplatform.android.repository.LocationOfInterestRepository
+import org.groundplatform.android.repository.SurveyRepository
+import org.groundplatform.android.ui.common.LocationOfInterestHelper
+
+/**
+ * Extracted “engine” that contains all mutable logic previously living in DataCollectionViewModel.
+ * - Keeps API surface similar so VM can delegate gradually.
+ * - Owns flows, sequencing, validation, and draft persistence.
+ *
+ * Usage: helper.initialize(savedStateHandle, viewModelScope) expose helper.uiState, jobName,
+ * loiName from VM delegate calls: onNextClicked(), onPreviousClicked(), saveCurrentState(), etc.
+ */
+class DataCollectionHelper
+@Inject
+constructor(
+  private val locationOfInterestHelper: LocationOfInterestHelper,
+  private val locationOfInterestRepository: LocationOfInterestRepository,
+  private val surveyRepository: SurveyRepository,
+) {
+
+  private lateinit var savedStateHandle: SavedStateHandle
+
+  private lateinit var jobId: String
+  private var loiId: String? = null
+  private val taskDataHandler = TaskDataHandler()
+  private lateinit var taskSequenceHandler: TaskSequenceHandler
+
+  fun initialize(savedStateHandle: SavedStateHandle) {
+    this.savedStateHandle = savedStateHandle
+    jobId = requireNotNull(savedStateHandle[TASK_JOB_ID_KEY])
+    loiId = savedStateHandle[TASK_LOI_ID_KEY]
+  }
+
+  suspend fun initialStateProvider(): DataCollectionUiState {
+    val survey =
+      loadSurveyOrReturnError()
+        ?: return DataCollectionUiState.Error("Survey failed to load. Please try again.")
+    val job =
+      resolveJobOrReturnError(survey) ?: return DataCollectionUiState.Error("Invalid jobId: $jobId")
+
+    val tasks = pickTasks(job, loiId)
+    if (!initSequenceOrReturnError(tasks)) {
+      return DataCollectionUiState.Error("No valid tasks for this job.")
+    }
+
+    val initialTaskId = computeInitialTaskId()
+    savedStateHandle[TASK_POSITION_ID] = initialTaskId
+
+    val loiName = computeLoiName(survey.id, loiId)
```

Comment:
I would suggest adding a try-catch block for the entire logic and return `Error` state with the thrown error. Then each of the methods can throw error with appropriate message internally. Also, instead of return error string message, let's return error_code or enum which can then be translated to proper UI strings in the ui layer.